tests/frontend/test_logger.py

```diff
@@ -0,0 +1,168 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Unit tests for the logging mechanism used in Strawberry Fields.
+
+The implementation of these unit tests is based on the solution used for testing logging in
+the Flask web application framework:
+https://github.com/pallets/flask/blob/master/tests/test_logging.py
+"""
+
+import logging
+
+import pytest
+
+import strawberryfields.api.job as job
+import strawberryfields.api.connection as connection
+import strawberryfields.engine as engine
+
+from strawberryfields.logger import logging_handler_defined, default_handler, create_logger
+
+modules_contain_logging = [job, connection, engine]
+
+@pytest.fixture(autouse=True)
+def reset_logging(pytestconfig):
+    """Reset the logging specific configurations such as handlers or levels as
+    well as manage pytest's LoggingPlugin."""
+    root_handlers = logging.root.handlers[:]
+    logging.root.handlers = []
+    root_level = logging.root.level
+
+    logging_plugin = pytestconfig.pluginmanager.unregister(name="logging-plugin")
+
+    yield
+
+    logging.root.handlers[:] = root_handlers
+    logging.root.setLevel(root_level)
+
+    if logging_plugin:
+        pytestconfig.pluginmanager.register(logging_plugin, "logging-plugin")
+
+
+@pytest.fixture(autouse=True)
+def reset_logging_module():
+    """Reset the logging specific configurations such as handlers or levels for
+    the module specific loggers."""
+    for module in modules_contain_logging:
+        logger = logging.getLogger(module.__name__)
+        logger.handlers = []
+        logger.setLevel(logging.NOTSET)
+
+
+@pytest.mark.parametrize("module", modules_contain_logging)
+class TestLogger:
+    """Tests for the functions that are used to create a logger"""
+
+    def test_logging_handler_defined(self, module):
+        """Tests that the logging_handler_defined function works correctly in
+        the following cases:
+
+        1. When a custom logger was just created and by default has no level
+           handler
+        2. Adding a handler to the root logger affects the custom logger
+        3. When propagation is set to False, only the handlers of the custom
+        logger are checked"""
+        logger = logging.getLogger(module.__name__)
+        assert not logging_handler_defined(logger)
+
+        handler = logging.StreamHandler()
+        logging.root.addHandler(handler)
+        assert logging_handler_defined(logger)
+
+        logger.propagate = False
+        assert not logging_handler_defined(logger)
+        logger.propagate = True
```

Comment:
Is there a purpose to this line? :thinking:

Reply:
No, thank you so much for catching that!